flag.go

```diff
@@ -991,13 +1004,13 @@ func (f *FlagSet) parseLongArg(s string, args []string, fn parseFunc) (a []strin
 	if len(split) == 2 {
 		// '--flag=arg'
 		value = split[1]
-	} else if flag.NoOptDefVal != "" {
-		// '--flag' (arg was optional)
-		value = flag.NoOptDefVal
 	} else if len(a) > 0 {
 		// '--flag arg'
 		value = a[0]
 		a = a[1:]
+	} else if flag.NoOptDefVal != "" {
+		// '--flag' (arg was optional)
+		value = flag.NoOptDefVal
 	} else {
```

Comment:
I haven't looked exactly _why_, but it looks like  this patch can cause a regression;  I  tried  using the `NoOptDefVal` on  a flag (`--pull`), and it now failed  when the flag  before it is a boolean;

```bash
docker run -it --rm --pull=never alpine

invalid argument "--pull=never" for "--rm" flag: strconv.ParseBool: parsing "--pull=never": invalid syntax
```

Maybe because boolean flags have an  implicit `NoOptDefVal` ?

Reply:
>  I tried using the NoOptDefVal on a flag (--pull), and it now failed when the flag before it is a boolean;

Actually; same happens without  that option set on the `--pull` flag, and I see the same for  other flags;

```bash
docker run -it --rm --label foo=bar alpine
invalid argument "--label" for "--rm" flag: strconv.ParseBool: parsing "--label": invalid syntax
```